sw/applications/example_iffifo/main.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright EPFL contributors.
+ * Licensed under the Apache License, Version 2.0, see LICENSE for details.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Author: Pierre Guillod <pierre.guillod@epfl.ch>
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "core_v_mini_mcu.h"
+
+#include "x-heep.h"
+#include "iffifo_regs.h"
+
+#include "mmio.h"
+#include "handler.h"
+#include "csr.h"
+#include "hart.h"
+
+#include "rv_plic.h"
+
+#include "dma.h"
+#include "dma_regs.h"
+#include "fast_intr_ctrl.h"
+
+
+#define PRINTF(fmt, ...)    printf(fmt, ## __VA_ARGS__)
+
+unsigned int IFFIFO_START_ADDRESS = EXT_PERIPHERAL_START_ADDRESS + 0x2000;
+
+int32_t to_fifo  [6]   __attribute__ ((aligned (4)))  = { 1, 2, 3, 4, 5, 6 };
+int32_t from_fifo[4]   __attribute__ ((aligned (4)))  = { 0, 0, 0, 0 };
+
+int8_t dma_intr_flag = 0;
+void dma_intr_handler_trans_done()
+{
+  dma_intr_flag = 1;
+}
+
+void protected_wait_for_dma_interrupt(void)
+{
+  while(!dma_is_ready()) {
+    CSR_CLEAR_BITS(CSR_REG_MSTATUS, 0x8);
+    if (!dma_is_ready()) {
+        wait_for_interrupt();
+    }
+    CSR_SET_BITS(CSR_REG_MSTATUS, 0x8);
+  }
+}
+
+iffifo_intr_flag = 0;
+static void handler_irq_iffifo( uint32_t int_id )
+{
+  mmio_region_t iffifo_base_addr = mmio_region_from_addr((uintptr_t)IFFIFO_START_ADDRESS);
+  mmio_region_write32(iffifo_base_addr, IFFIFO_INTERRUPTS_REG_OFFSET, 0b0);
+  iffifo_intr_flag = 1;
+  PRINTF(" ** REACH intr. fired.\n");
+}
+
+static dma_target_t tgt_src;
+static dma_target_t tgt_dst;
+static dma_trans_t trans;
+
+int compare_print_fifo_array(void) {
+  int errors = 0;
+  PRINTF("from_fifo = {");
+  for (int i = 0; i < 4; i+=1) {
+    PRINTF("%d",from_fifo[i]);
+    if(i != 4-1) {PRINTF(", ");};
+    if (to_fifo[i]+1 != from_fifo[i]) {++errors;}
+  }
+  PRINTF("}\n");
+  return errors;
+}
+
+void print_status_register(void)
+{
+  mmio_region_t iffifo_base_addr = mmio_region_from_addr((uintptr_t)IFFIFO_START_ADDRESS);
+  int32_t status = mmio_region_read32(iffifo_base_addr, IFFIFO_STATUS_REG_OFFSET);
+  PRINTF("STATUS = ");
+  PRINTF(status & 1 ? "E" : "-"); // FIFO empty
```

Comment:
pls use the following defines avilable in the regs.h 

#define IFFIFO_STATUS_EMPTY_BIT 0
#define IFFIFO_STATUS_AVAILABLE_BIT 1
#define IFFIFO_STATUS_REACHED_BIT 2
#define IFFIFO_STATUS_FULL_BIT 3